libcudacxx/include/cuda/__mdspan/host_device_accessor.h

```diff
@@ -0,0 +1,283 @@
+  static constexpr bool __is_offset_noexcept    = noexcept(_Accessor{}.offset(__data_handle_type{}, 0));
+
+  static_assert(!is_host_device_managed_accessor_v<_Accessor>,
+                "cuda::host_accessor/cuda::device_accessor/cuda::managed_accessor accessor cannot be nested");
+
+  _LIBCUDACXX_HIDE_FROM_ABI static constexpr void __check_device_pointer(__data_handle_type __p)
+  {
+    _CCCL_ASSERT(__is_device_pointer(__p), "cuda::device_accessor data handle is not a DEVICE pointer");
+  }
+
+  template <typename _Sp = bool> // lazy evaluation
+  _LIBCUDACXX_HIDE_FROM_ABI static constexpr void __check_host_path() noexcept
+  {
+    static_assert(sizeof(_Sp) != sizeof(_Sp), "cuda::device_accessor cannot be used in HOST code");
+  }
+
+public:
+  using offset_policy = device_accessor;
+
+  _CCCL_HIDE_FROM_ABI device_accessor() noexcept(__is_ctor_noexcept) = default;
+
+  _CCCL_HIDE_FROM_ABI device_accessor(const device_accessor&) noexcept(__is_copy_ctor_noexcept) = default;
+
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr __reference access(__data_handle_type __p, size_t __i) const
+    noexcept(__is_access_noexcept)
+  {
+    NV_IF_ELSE_TARGET(NV_IS_DEVICE, (__check_device_pointer(__p);), (__check_host_path();))
+    return _Accessor::access(__p, __i);
+  }
+
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr __data_handle_type offset(__data_handle_type __p, size_t __i) const
+    noexcept(__is_offset_noexcept)
+  {
+    NV_IF_ELSE_TARGET(NV_IS_DEVICE, (__check_device_pointer(__p);), (__check_host_path();))
+    return _Accessor::offset(__p, __i);
+  }
+};
+
+/***********************************************************************************************************************
+ * Managed Accessor
+ **********************************************************************************************************************/
+
+template <typename _Accessor>
+struct managed_accessor : public _Accessor
+{
+private:
+  using __data_handle_type = typename _Accessor::data_handle_type;
+  using __reference        = typename _Accessor::reference;
+
+  static constexpr bool __is_ctor_noexcept      = noexcept(_Accessor{});
+  static constexpr bool __is_copy_ctor_noexcept = noexcept(_Accessor{});
+  static constexpr bool __is_access_noexcept    = noexcept(_Accessor{}.access(__data_handle_type{}, 0));
+  static constexpr bool __is_offset_noexcept    = noexcept(_Accessor{}.offset(__data_handle_type{}, 0));
+
+  static_assert(!is_host_device_managed_accessor_v<_Accessor>,
+                "cuda::host_accessor/cuda::device_accessor/cuda::managed_accessor accessor cannot be nested");
+
+  template <typename __data_handle_type>
+  _LIBCUDACXX_HIDE_FROM_ABI static constexpr bool __is_managed_pointer(__data_handle_type __p)
+  {
+    if constexpr (::cuda::std::is_pointer_v<__data_handle_type>)
+    {
+      cudaPointerAttributes __attrib;
+      auto __status = ::cudaPointerGetAttributes(&__attrib, __p);
+      return __status == ::cudaSuccess
```

Comment:
If managed accessor means that the memory can be derefenced on the host and device this check is incorrect.
All host side allocations ( e.g. malloc ) will return true since they are `cudaMemoryTypeUnregistered`.

What we want is a validity check of both the host and device ptr. If both are true we know the allocation is safe to use ( this captures power9 / G+H / openrmm  allocs ) .
So roughly like:
```
cudaError_t err = cudaPointerGetAttributes(&pointer_attr, p);
retun pointer_attr.devicePointer != nullptr && pointer_attr.hostPointer ==  pointer_attr.devicePointer